src/buffer/out/Row.cpp

```diff
@@ -191,8 +347,331 @@ OutputCellIterator ROW::WriteCells(OutputCellIterator it, const til::CoordType i
+    // * colEnd is now (colBeg, _columnCount]
+
+    // Algorithm explanation
+    //
+    // Task:
+    //   Replace the characters in cells [colBeg, colEnd) with a single `width`-wide glyph consisting of `chars`.
+    //
+    // Problem:
+    //   Imagine that we have the following ROW contents:
+    //     "xxyyzz"
+    //   xx, yy, zz are 2 cell wide glyphs. We want to insert a 2 cell wide glyph ww at colBeg 1:
+    //       ^^
+    //       ww
+    //   An incorrect result would be:
+    //     "xwwyzz"
+    //   The half cut off x and y glyph wouldn't make much sense, so we need to fill them with whitespace:
+    //     " ww zz"
+    //
+    // Solution:
+    //   Given the range we want to replace [colBeg, colEnd), we "extend" it to encompass leading (preceding)
+    //   and trailing wide glyphs we partially overwrite resulting in the range [colExtBeg, colExtEnd), where
+    //   colExtBeg <= colBeg and colExtEnd >= colEnd. In other words, the to be replaced range has been "extended".
+    //   The amount of leading whitespace we need to insert is thus colBeg - colExtBeg
+    //   and the amount of trailing whitespace colExtEnd - colEnd.
+
+    // Extend range downwards (leading whitespace)
+    uint16_t colExtBeg = colBeg;
+    // Safety: colExtBeg is [0, _columnCount], because colBeg is.
+    const uint16_t chExtBeg = _uncheckedCharOffset(colExtBeg);
+    // Safety: colExtBeg remains [0, _columnCount] due to colExtBeg != 0.
+    for (; colExtBeg != 0 && _uncheckedIsTrailer(colExtBeg); --colExtBeg)
+    {
+    }
+
+    // Extend range upwards (trailing whitespace)
+    uint16_t colExtEnd = colEnd;
+    // Safety: colExtEnd cannot be incremented past _columnCount, because the last
+    // _charOffset at index _columnCount will never get the CharOffsetsTrailer flag.
+    for (; _uncheckedIsTrailer(colExtEnd); ++colExtEnd)
+    {
+    }
+    // Safety: After the previous loop colExtEnd is [0, _columnCount].
+    const uint16_t chExtEnd = _uncheckedCharOffset(colExtEnd);
+
+    const uint16_t leadingSpaces = colBeg - colExtBeg;
+    const uint16_t trailingSpaces = colExtEnd - colEnd;
+    const size_t chExtEndNew = chars.size() + leadingSpaces + trailingSpaces + chExtBeg;
+
+    if (chExtEndNew != chExtEnd)
+    {
+        _resizeChars(colExtEnd, chExtBeg, chExtEnd, chExtEndNew);
+    }
+
+    // Add leading/trailing whitespace and copy chars
+    {
+        auto it = _chars.begin() + chExtBeg;
+        it = fill_n_small(it, leadingSpaces, L' ');
+        it = copy_n_small(chars.begin(), chars.size(), it);
+        it = fill_n_small(it, trailingSpaces, L' ');
+    }
+    // Update char offsets with leading/trailing whitespace and the chars columns.
+    {
+        auto chPos = chExtBeg;
+        auto it = _charOffsets.begin() + colExtBeg;
+
+        it = iota_n_mut(it, leadingSpaces, chPos);
+
+        *it++ = chPos;
+        it = fill_small(it, _charOffsets.begin() + colEnd, gsl::narrow_cast<uint16_t>(chPos | CharOffsetsTrailer));

```

Comment:
I forget why we store a Trailer flag and don't just determine it by looking to see if the column/offset value was repeated.

Reply:
The trailer flag allows us to skip looking at neighboring cells to determine whether the offset was repeated. It's basically for convenience to make writing the forward/backward loops easier to write.